### ISSUE TITLE
Js files?

### ISSUE THREAD

**Issue by bertvh** [500tech/angular2-tree-component#1]
The npm package does not contain the js file(s) of this component. This means the component can't be loaded with system.js. Do you have any plans to fix this (or am I missing something here)?

Thanks.

**Comment by adamkleingit** [comment]
Hey,
I haven't tested it with SystemJS yet - and yes I want to support it obviously.

Can you share with me the code that is trying to use the library so I can look into it?
Thanks

**Comment by bertvh** [comment]
I installed the treeview dependency using npm install so it's in my node_modules.

Then, I write something like this to use it in my components:

    import {TreeComponent} from "angular2-tree-component"

from which the Typescript compiler generates statements like this:

    require('angular2-tree-component')

There are picked up by SystemJs, which tried to load `http://.../<node_modules>/angular2-tree-component.js` (not found of course) <- this must be a js file since it's a lib and hence not being compiled.

**Comment by adamkleingit** [comment]
So to fix this I need to compile the entire lib into a single
angular2-tree-component.js file and include it in the root folder of the
GIT repo?

-- 
- Adam Klein <http://il.linkedin.com/in/adamklein500>
- CTO @ 500Tech.com <http://500tech.com/>
- AngularJS, ReactJS, NodeJS
- Development, consulting, training and public speaking
- (+972) 52 747 5633

**Comment by bertvh** [comment]
Ideally, you would let Grunt or Gulp or whatever compile everything and publish to npm but just checking in the compiled result should work (and is probably easier to set up).

You could look at the ng2-bootstrap module

- Their npm package with JS files: https://www.npmjs.com/package/ng2-bootstrap
- Their Github without JS files: https://github.com/valor-software/ng2-bootstrap

**Comment by isaacplmann** [comment]
Let me know if you want help setting this up.  I'm using this component in a typescript environment, but it throws a bunch of tslint errors (since you're not using the same tslint rules we are).  It compiles fine, but it would be nice to not have to wade through meaningless errors looking for real ones.

**Comment by 500tech-user** [comment]
Sure, can you send me your tslint configuration?

BTW, I'd really appreciate your opinions as early users of the package, do
you find it useful, what features are missing, etc.

Thanks!

On Wed, May 25, 2016 at 1:35 PM, isaacplmann <notifications@github.com>
wrote:

> Let me know if you want help setting this up. I'm using this component in
> a typescript environment, but it throws a bunch of tslint errors (since
> you're not using the same tslint rules we are). It compiles fine, but it
> would be nice to not have to wade through meaningless errors looking for
> real ones.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-221699137>
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by isaacplmann** [comment]
Actually, I misspoke.  The warnings I’m getting are from the tsc compiler itself.

10 or so errors of this type:

node_modules/angular2-tree-component/lib/models/tree.model.ts:46:14     Property 'last' does not exist on type '{}'.


General feedback:

1. Nicely documented and structured
2. The keyboard navigation is a great feature
3. The only missing features that I needed were double click and context menu events, but I implemented those in the pull request.

**Comment by isaacplmann** [comment]
Here are a few other features that I’ve implemented outside of the component, which might be nice to include in the component itself:

1. Optionally use a provided nodesAreEqual function instead of a simple id field (some of my data objects have multiple keys)
2. List of initially expanded nodes
3. A canSelectFilter function to determine if a node is selectable - unselectable nodes call toggle() instead of toggleActivated()
4. A canShowFilter function to determine if a node is displayed in the tree.  If a node has children it is displayed regardless of the canShowFilter.

I could incorporate any or all of these features into a pull request if you think they’re worth adding in.

Isaac

**Comment by neelsbr** [comment]
Can anyone tell me what is the final call on this? How to add in system.config.js file?

**Comment by 500tech-user** [comment]
Hey,
I was on a long vacation until now, and will try to solve the system.js
issue ASAP.
If anyone else already managed to solve it in a good way I'll be happy to
merge it.

Cheers

On Mon, Jun 13, 2016 at 6:50 PM, neelsbr <notifications@github.com> wrote:

> Can anyone tell me what is the final call on this? How to add in
> system.config.js file?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225623022>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACUsQ6uR4WwMP_suZT81rnm5jemZUKCfks5qLXwqgaJpZM4Iklc7>
> .
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by isaacplmann** [comment]
I have some time this morning to implement saving the compiled js to npm.  I’d use the angular2-library-seed template (https://github.com/preboot/angular2-library-seed <https://github.com/preboot/angular2-library-seed>) to have webpack make the js files.  If you’re not comfortable with webpack though, you may want to choose a different solution.  Let me know if you’d like me to move forward with it.

Isaac

**Comment by 500tech-user** [comment]
Webpack is great. Thanks!

On Tue, Jun 14, 2016 at 4:09 PM, isaacplmann <notifications@github.com>
wrote:

> I have some time this morning to implement saving the compiled js to npm.
> I’d use the angular2-library-seed template (
> https://github.com/preboot/angular2-library-seed <
> https://github.com/preboot/angular2-library-seed>) to have webpack make
> the js files. If you’re not comfortable with webpack though, you may want
> to choose a different solution. Let me know if you’d like me to move
> forward with it.
>
> Isaac
>
>
> > On Jun 14, 2016, at 5:28 AM, 500Tech <notifications@github.com> wrote:
> >
> > Hey,
> > I was on a long vacation until now, and will try to solve the system.js
> > issue ASAP.
> > If anyone else already managed to solve it in a good way I'll be happy to
> > merge it.
> >
> > Cheers
> >
> > On Mon, Jun 13, 2016 at 6:50 PM, neelsbr <notifications@github.com>
> wrote:
> >
> > > Can anyone tell me what is the final call on this? How to add in
> > > system.config.js file?
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225623022
> >,
> > > or mute the thread
> > > <
> https://github.com/notifications/unsubscribe/ACUsQ6uR4WwMP_suZT81rnm5jemZUKCfks5qLXwqgaJpZM4Iklc7
> >
> > > .
> > >
> >
> >
> >
> > --
> > *Adam Klein*
> > *CTO & Co-Founder*
> > *Tel.* +972-52-747-5633
> > *blog.500tech.com <http://blog.500tech.com>*
> > meetup.com/angularjs-il
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub <
> https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225827793>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/AA0lQFgLDgpt3Lo-daog0HD3QGqgIanOks5qLnQggaJpZM4Iklc7
> >.
>
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225875610>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACUsQ_i8oeAwC9XhdQm-woDyHiTfyjT_ks5qLqgKgaJpZM4Iklc7>
> .
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by isaacplmann** [comment]
I started to use angular2-library-seed, but it ended up being a lot more than I needed.  So I just set up npm scripts that use the typescript compiler.  Workflow goes like this:

1. npm install
2. npm run build
3. Bump version numbers in package.json
4. npm publish

*.js and *.d.ts files are ignored by git (except lib/defs/angular2-tree.d.ts).
*.ts files are ignored by npm

**Comment by 500tech-user** [comment]
So .ts files will not be pulled when npm installing the library?


On Tue, Jun 14, 2016 at 10:40 PM, isaacplmann <notifications@github.com>
wrote:

> I started to use angular2-library-seed, but it ended up being a lot more
> than I needed. So I just set up npm scripts that use the typescript
> compiler. Workflow goes like this:
>
>    1. npm install
>    2. npm run build
>    3. Bump version numbers in package.json
>    4. npm publish
>
> *.js and *.d.ts files are ignored by git (except
> lib/defs/angular2-tree.d.ts).
> *.ts files are ignored by npm
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225993185>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACUsQ_FKTh-YLYCzvTlNPail4WkpK_yBks5qLwO6gaJpZM4Iklc7>
> .
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by isaacplmann** [comment]
Correct. The ts files were what was giving me all those errors when I
compile my app.

You could theoretically include the ts files too but I have yet to see
another library that includes them.

**Comment by 500tech-user** [comment]
OK I really hope to have time to merge this tonight.
Thanks

On Wed, Jun 15, 2016 at 12:33 PM, isaacplmann <notifications@github.com>
wrote:

> Correct. The ts files were what was giving me all those errors when I
> compile my app.
>
> You could theoretically include the ts files too but I have yet to see
> another library that includes them.
>
> On Wednesday, June 15, 2016, 500Tech <notifications@github.com> wrote:
>
> > So .ts files will not be pulled when npm installing the library?
> >
> >
> > On Tue, Jun 14, 2016 at 10:40 PM, isaacplmann <notifications@github.com
> > <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> > wrote:
> >
> > > I started to use angular2-library-seed, but it ended up being a lot
> more
> > > than I needed. So I just set up npm scripts that use the typescript
> > > compiler. Workflow goes like this:
> > >
> > > 1. npm install
> > > 2. npm run build
> > > 3. Bump version numbers in package.json
> > > 4. npm publish
> > >
> > > *.js and *.d.ts files are ignored by git (except
> > > lib/defs/angular2-tree.d.ts).
> > > *.ts files are ignored by npm
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> >
> https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-225993185
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe/ACUsQ_FKTh-YLYCzvTlNPail4WkpK_yBks5qLwO6gaJpZM4Iklc7
> > >
> > > .
> > >
> >
> >
> >
> > --
> > *Adam Klein*
> > *CTO & Co-Founder*
> > *Tel.* +972-52-747-5633
> > *blog.500tech.com <http://blog.500tech.com>*
> > meetup.com/angularjs-il
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-226102933
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AA0lQLIwSQV1OHnTffC-qVNDsRoUP5Fzks5qL6DogaJpZM4Iklc7
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/1#issuecomment-226136592>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACUsQxrKVykrkMcuEbFxJsNG0P6Fd_P-ks5qL8bMgaJpZM4Iklc7>
> .
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by isaacplmann** [comment]
There are only a handful of files that were changed to get the compile to
js feature to work. You could just copy over the last commit if you're
uncertain about the rest of my changes.

**Comment by adamkleingit** [comment]
Hey Isaac, I merged your pull request #2 and made a few changes / fixes.
I'm compiling everything to 'dist' folder, and defined angular2 as peerDependency.

I added 2 example folders, one with webpack and one with systemjs, both using angular2-tree-component successfully.

I had one problem though with systemjs - I had to manually install lodash as a consumer of angular2-tree-component, otherwise systemjs couldn't resolve it.

Anyone knows what's the right way to deal with library dependencies in typescript + npm + systemjs?

Maybe it's because lodash is defined as ambient in typings.json file?
